Add errorbars to mplot3d

vlas-sokolov opened the pull request:
Motivated by the 'new-contributor-friendly' and 'wishlist' tags in #568, and having once asked [this](http://stackoverflow.com/questions/26245442/adding-errorbars-to-3d-plot-in-matplotlib) question on SO, this PR mimics the errorbar functionality present in the standard lib into `mplot3d`.

For a demonstration, I have rewritten a short snippet from [here](
https://github.com/vlas-sokolov/matplotlib/tree/errorbars-mplot3d) into and [put in on Gist](
https://gist.github.com/vlas-sokolov/9987d136a0c68abdf5c3a0d9c196822c). It generates the figures below:

![z-axis_errors](https://cloud.githubusercontent.com/assets/11835115/22652227/f8fcebfa-ec85-11e6-993f-f41caf5e7049.png)
![x-axis_errors](https://cloud.githubusercontent.com/assets/11835115/22652229/faec3d3a-ec85-11e6-9c16-0cab356562a0.png)

The new function was rewritten from its 2D counterpart. I aimed to follow the old structure as close as possible, and I think that most of the optional keywords have been successfully ported.

One issue that I'd like to get input in is the use of markers in error bar caps and (especially) limit symbols. The markers tend to rotate along with the camera position, and a 180 degree rotation along one of the axes may flip the markers into a rather confusing plot. What should be done with this? Is there a framework for projecting the markers in mplot3d?

As I'm a new contributor and might have missed some subtleties in the implementation, any feedback is highly appreciated!

WeatherGod commented:
(sent this earlier, but it hasn't shown up yet, so I am posting this again)

Heads up. There is another PR that is adding a new feature to errorbar that might need to be considered here, too: #6280

This looks neat. The marker problem is difficult to handle correctly and there isn't any framework for dealing with it partly because mplot3d does not properly integrate into the overall transforms system. Markers are actually avoided in Axis3D for ticks partly because of the difficulty in orienting and sizing them correctly. Now, it isn't impossible to get them right. Depending on the marker, they are implemented as PolyCollection or PatchCollection, if I remember correctly. Both of these have 3D analogs as Poly3DCollection and Patch3DCollection. In `art3d.py`, there are `patch_collection_2d_to_3d()` and `poly_collection_2d_to_3d` that will convert the 2D instances into 3D instances in-place. That *might* work for you here.

vlas-sokolov commented:
@WeatherGod I implemented (read: copied over from #6280) the tuple offsets for `errorevery`. Seems to work, although the overlap of errorbars is mainly set by the viewing angle. Still, for `ax.view_init(10, 90)` and a setup from #6280 the 3d equivalent is as follows:

![errorevery_3d](https://cloud.githubusercontent.com/assets/11835115/22702824/4b1cf65a-ed62-11e6-9ea3-1fd4b4e5f098.png)

I briefly looked over the marker issue again, but I can't find a way to improve it with the `2d_to_3d` functions in `art3d.py`. Is it worth keeping the markers? In principle, we can ditch them in favor of 3d lines. The limits, etc. won't look as pretty though.

vlas-sokolov commented:
Not sure why Travis fails, could someone give a tip on how to fix this? Seems unrelated to me.

anntzer commented:
@vlas-sokolov Let's figure out what to do with https://github.com/matplotlib/matplotlib/pull/6280#pullrequestreview-21158726 before including the extended version of `errorevery`.

vlas-sokolov commented:
@anntzer all right then, standing by.

dstansby commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2707,6 +2708,299 @@ def calc_arrow(uvw, angle=15):
+
+        Parameters
+        ----------
+        x : scalar
```
I may be missing something here, but should `errorbar` be able to take an array of data points to plot more than one bar at a time?

vlas-sokolov replied:
Good point. I took and modified the docstring from the original 2d function in `axes/_axes.py`, and it was set there in this way. Seems to have been introduced here: https://github.com/matplotlib/matplotlib/commit/4d2c994a5c62214b6728447dc137534410d00f53

dstansby commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2707,6 +2708,299 @@ def calc_arrow(uvw, angle=15):
+        y : scalar
+        z : scalar
+
+        xerr/yerr/zerr : scalar or array-like, shape(n,1) or shape(2,n).
```
Needs a note that `xerr/yerr/zerr` are optional

vlas-sokolov replied:
`err`-like arguments are optional to allow flexibility in choosing the dimension your errors are plotten in. For example, one can specify `xerr` and `zerr` to be `None`, so that only y-axis errors would be plotted. Therefore, any given argument from the `xerr/yerr/zerr` can be optional.

tacaswell commented:
I completely missed that this PR came in, cool work! 

What is the current state of this?

vlas-sokolov commented:
@tacaswell all the basic functionality should be there already. There is, however, an issue with using the markers in limit symbols - I didn't figure out how to rotate them in 3d properly, so certain projections look wrong for the marker caps. Perhaps it is wiser to draw the caps / limits with lines instead.

Also, following @WeatherGod's suggestion to add `errorevery` argument - is it better to roll it back for now until #6280 is resolved?

vlas-sokolov commented:
I'm picking this up again.

I've rebased against current master and fixed a number of issues, including adapting the old PR code to the changes in errorbar's 2D version over the last three years.

While the overall functionality is there, I see several remaining issues (WIP!):

- [ ] When both upper/lower limits set, the errorbars don't drawn (only caps do)
- [ ] Legend appears multiple times if more than one coordinate has error bars
- [ ] Allow for all the input shapes in error data
- [ ] Cap rotation is (still) broken in 3d

tacaswell commented:
@vlas-sokolov Awesome!  

attn @WeatherGod and @eric-wieser

eric-wieser commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2770,6 +2771,338 @@ def permutation_matrices(n):
+            _lomask = lomask | ((lomask == himask) & everymask)
+            _himask = himask | ((lomask == himask) & everymask)
+
+            lows = [d - e if m else d for d, e, m in zip(data, err, _lomask)]
```
```suggestion
            lows = np where(_lomask, data - err, data)
```

vlas-sokolov replied:
@eric-wieser the 2d function I've adapted this from makes a big fuss about not using arrays for data (e.g. `xywhere`, the [equivalent of this function](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/axes/_axes.py#L3372-L3381)), so I tried to use numpy arrays for masking only. If we can ditch that and just use arrays, it would simplify things a lot, but I'm not sure that's the case.

eric-wieser replied:
That's good to know, thanks. I know a lot of work recently went into trying to make `mplot3d` faster by using arrays more, so it would be good to know why the 2d one avoids it.

vlas-sokolov replied:
Now that I look at it, there's a FIXME item down the file where I've actually converted everything into arrays. I'll add fixing that to my todo-list for the PR.

vlas-sokolov replied:
Meanwhile, if anyone knows of a testable usecase where using arrays is not ok, I'd like to see it (for testing reasons).

vlas-sokolov replied:
@eric-wieser I don't know anymore why the 2d version does this the way it does.  The original code comes from pre-2007 commits (as far as git blame took me), and I don't even know those comments are relevant, because
1. I forced np.array on errors in 2d errorbar and ran `pytest lib/matplotlib/tests/test_axes.py` and `pytest lib/matplotlib/tests/test_units.py` successfully,
2. the only other place I've found units being an issue (https://github.com/matplotlib/matplotlib/pull/10118) was closed with only upstream changes, and
3. Some special cases, such as plotting pandas dataframes, appear to work normally regardless of forcing arrays

Given that I'm already using arrays for errors elsewhere in the code, can I suggest I implement the change here and the unit business gets resolved in a separate issue (as I don't seem to understand it)?

eric-wieser commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2770,6 +2771,338 @@ def permutation_matrices(n):
+        had_data = self.has_data()
+
+        plot_line = (fmt.lower() != 'none')
+        label = kwargs.pop("label", None)
```
Instead of this, can you add `*, label=None` to the signature? Or is there a deliberate reason to not list this in the argument list?

vlas-sokolov replied:
Probably the reason is that 2d `ax.errorbar` was written for Python2, where that signature would throw a `SyntaxError`. I made this PR taking the following comment from #568 - "just mirror ax.errorbar as closely as possible" as a guideline (reviewing its source side-by-side might be easier to show what bits are actually new in this PR). Should I still change it?

eric-wieser commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2770,6 +2771,332 @@ def permutation_matrices(n):
+                lo_xyz = _mask_lists(x, y, z, upmask)
+                up_xyz = _mask_lists(x, y, z, lomask)
+
+                x0, y0, z0 = np.vstack([np.c_[lo_xyz],
+                                        np.c_[up_xyz]]).T
```
I'd argue this is clearer as
```suggestion
                x0, y0, z0 = np.concatenate([lo_xyz, up_xyz], axis-1)
```

vlas-sokolov replied:
Done in 159b6b5

eric-wieser commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2770,6 +2771,332 @@ def permutation_matrices(n):
+
+                x0, y0, z0 = np.vstack([np.c_[lo_xyz],
+                                        np.c_[up_xyz]]).T
+                dx, dy, dz = np.vstack([np.c_[lolims_xyz] - np.c_[lo_xyz],
+                                        np.c_[uplims_xyz] - np.c_[up_xyz]]).T
```
```suggestion
                dx, dy, dz = np.concatenate([
                    np.array(lolims_xyz) - np.array(lo_xyz),
                    np.array(uplims_xyz) - np.array(up_xyz)
                ], axis=-1)
```

vlas-sokolov replied:
Done in 159b6b5

eric-wieser commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2770,6 +2771,334 @@ def permutation_matrices(n):
+                (xlo, xup), (ylo, yup), (zlo, zup) = limits
+                lomask = lolims & everymask
+                upmask = uplims & everymask
+                lolims_xyz = _mask_lists(xlo, ylo, zlo, upmask)
+                uplims_xyz = _mask_lists(xup, yup, zup, lomask)
+                lo_xyz = _mask_lists(x, y, z, upmask)
+                up_xyz = _mask_lists(x, y, z, lomask)
+
+                x0, y0, z0 = x0, y0, z0 = np.concatenate([lo_xyz, up_xyz],
+                                                         axis=-1)
+                dx, dy, dz = np.concatenate([
+                    np.array(lolims_xyz) - np.array(lo_xyz),
+                    np.array(uplims_xyz) - np.array(up_xyz)], axis=-1)
```
May as well halve the number of times you create the same arrays:
```suggestion
                lolims_xyz = np.array(_mask_lists(xlo, ylo, zlo, upmask))
                uplims_xyz = np.array(_mask_lists(xup, yup, zup, lomask))
                lo_xyz = np.array(_mask_lists(x, y, z, upmask))
                up_xyz = np.array(_mask_lists(x, y, z, lomask))

                x0, y0, z0 = np.concatenate([lo_xyz, up_xyz], axis=-1)
                dx, dy, dz = np.concatenate([lolims_xyz - lo_xyz, uplims_xyz - up_xyz], axis=-1)
```

vlas-sokolov replied:
Thanks, that's a great suggestion. Fixed it in 5db6976.

vlas-sokolov commented:
I've finished addressing the remaining issues on the PR, and it's ready for the final review. Here's how the new feature looks at the moment:

```python
fig = plt.figure() 
ax = fig.gca(projection="3d") 
 
d = [1,2,3,4,5] 
e = [.5]*5 
f = [[.5]*5, [.2, .1, 1.1, 0.5, .2]] 
ax.errorbar(x=d, y=d, z=d, xerr=f, yerr=e, zerr=e, capsize=3,
            zuplims=[False, True, False, True, True], 
            zlolims=[True, False, False, True, False], 
            yuplims=True, 
            arrow_length_ratio=0.6, 
            ecolor='purple', label='Error lines') 
ax.legend()                                                  

```

![errorbars3d](https://user-images.githubusercontent.com/11835115/81833100-94ca8500-953f-11ea-8143-da8a233812aa.png)

eric-wieser commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2770,6 +2771,335 @@ def permutation_matrices(n):
+            # for compatibility with the 2d errorbar function, when both upper
+            # and lower limits specified, we need to draw the markers / line -
+            # whether or not using both limits makes any sense (it doesn't)
+            _lomask = lomask | ((lomask == himask) & everymask)
+            _himask = himask | ((lomask == himask) & everymask)
```
No point computing the same mask twice:
```suggestion
            common_mask = (lomask == himask) & everymask
            _lomask = lomask | common_mask
            _himask = himask | common_mask
```
feel free to pick a better name.

vlas-sokolov commented:
Can someone let me know if I can do anything to make the coveted "Approved" check mark happen sooner?

eric-wieser commented on examples/mplot3d/errorbar3d.py:
```diff
@@ -0,0 +1,31 @@
+x, y, z = np.sin(t), np.cos(3*t), np.sin(5*t)
+
+estep = 15
+zuplims = [True if (not i % estep and i // estep % 3 == 0)
+           else False for i in range(t.size)]
+zlolims = [True if (not i % estep and i // estep % 3 == 2)
+           else False for i in range(t.size)]
```
Perhaps easier to read as numpy arrays:
```suggestion
i = np.arange(t.size)
zuplims = (i % estep == 0) & (i // estep % 3 == 0)
zlolims = (i % estep == 0) & (i // estep % 3 == 2)
```

vlas-sokolov commented:
I've fixed the merge conflict against the latest master branch.

QuLogic commented:
The 2D `errorbar` was simplified somewhat by #15037; is there anything there that can/should be applied here for the 3D case?

vlas-sokolov commented:
> The 2D `errorbar` was simplified somewhat by #15037; is there anything there that can/should be applied here for the 3D case?

@QuLogic I wasn't aware of it. Will try to adapt the 3d code accordingly for the relevant bits.

vlas-sokolov commented:
@QuLogic I've ported most of the changes over to the 3d version.

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2913,6 +2915,341 @@ def permutation_matrices(n):
+
+            # for compatibility with the 2d errorbar function, when both upper
+            # and lower limits specified, we need to draw the markers / line -
+            # whether or not using both limits makes any sense (it doesn't)
```
What doesn't make sense exactly?

vlas-sokolov replied:
Upper limits and lower limits should technically be mutually exclusive on a data point - the true value of it is either below the arrow (upper limit) or above it (lower limit). It doesn't make much sense for both to be specified at the same time.

However, the 2d errorbar API allows for it - so the code block here replicates that functionality.

QuLogic replied:
Do you mean from a physical standpoint? I'm not sure I understand why it doesn't make sense from the library point of view.

vlas-sokolov replied:
Maybe I felt grumpy when adding the comment - I'll remove that last bit.

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2913,6 +2915,341 @@ def permutation_matrices(n):
+                for i, coord in enumerate([x, y, z])]
+            (xl, xh), (yl, yh), (zl, zh) = coorderr
+
+            # draws capmarkers - flat caps othogonal to the error bars
```
```suggestion
            # draws capmarkers - flat caps orthogonal to the error bars
```

QuLogic commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2913,6 +2915,341 @@ def permutation_matrices(n):
+        minz, maxz = _digout_minmax(coorderrs, 'z')
+        self.auto_scale_xyz((minx, maxx), (miny, maxy), (minz, maxz), had_data)
+
+        # Adapting errorbar containers for 3d case, assuming z-axis points "up"
```
Would it maybe be better to create an `Errorbar3DContainer` instead?

vlas-sokolov replied:
I have tried subclassing `ErrorbarContainer` but it resulted in quite a few errors that I failed to fix. I would suggest postponing that, given that no other axes3d methods rely on containers yet.

QuLogic commented on lib/mpl_toolkits/tests/test_mplot3d.py:
```diff
@@ -975,6 +975,40 @@ def test_minor_ticks():
     ax.set_zticklabels(["half"], minor=True)
 
 
+@mpl3d_image_comparison(['errorbar3d_errorevery.png'], tol=0.03)
```
Is there a reason for the tolerance?

QuLogic replied:
If we don't need it, then it's good not have it.

QuLogic commented:
I would like that we figure out kwarg handling (#17930) before releasing this, but it doesn't have to block merging (I can rebase that and update 3D too.)

vlas-sokolov commented on lib/mpl_toolkits/mplot3d/axes3d.py:
```diff
@@ -2913,6 +2915,341 @@ def permutation_matrices(n):
+            label=label)
+        self.containers.append(errorbar_container)
+
+        return errlines, caplines, limmarks
```
Yes. The errorbar container here is used only to make the legends work properly. To make this container, I have to "pretend" the data is in 2d, so if we would return the `ErrorbarContainer`, then we'd lose all the information on the 3rd axis. As per my comment above, I simply could not get the 3D container to work and the containers aren't used in other places in axes3d.py, so I stick to returning a tuple of drawn elements.

vlas-sokolov commented:
@QuLogic if it's all right with you, I would prefer the latter option (merge, rebase, update the 3d code), since you're more familiar with the changes in #17930.

vlas-sokolov commented:
I'm not sure if the latest commit did anything to cause the tests to break - I removed the tolerance from errorbar tests but it doesn't seem the failures came from there?

jklymak commented:
No there was a pandas incompatibility.  Should be fixed if you rebase on master.

jklymak commented:
Ooops, no this isn't the pandas thing, but some other issue with CI, but not an issue w/ your PR....

QuLogic commented:
I restarted them; it seems they're fine.
